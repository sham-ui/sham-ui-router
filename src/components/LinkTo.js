import { inject, options } from 'sham-ui';
import { onclick, ref } from 'sham-ui-directives';
import LinkToComponentTemplate from './LinkTo.sht';

export default class LinkToComponent extends LinkToComponentTemplate {
    constructor( options ) {
        super( {
            ...options,
            directives: {
                onclick,
                ref,
                ...options.directives
            }
        } );
    }

    @inject router;
    @inject( 'sham-ui:store' ) componentStore;

    @options text = '';
    @options useActiveClass = false;
    @options activeClass = 'active';
    @options className = '';

    @options get params() {
        return {};
    }

    generateURL( path, params ) {
        return this.lastGeneratedURL = this.router.generate( path, params );
    }

    // eslint-disable-next-line no-unused-vars
    isActive( path, params ) {
        return this.lastGeneratedURL === this.router.lastRouteResolved().url;
    }

    generateClassName( className, useActiveClass, activeClass, path, params ) {
        return [
            className,
            useActiveClass && this.isActive( path, params ) ? activeClass : ''
        ].join( ' ' ).trim();
    }

    _registryType() {
        const activeLinkType = ACTIVE_PAGE_LINK_TYPE;
        const storeHasType = this.componentStore.byType.has( activeLinkType );
        if ( this.options.useActiveClass ) {
            if ( !storeHasType ) {
                this.componentStore.byType.set( activeLinkType, new Set() );
            }
            this.componentStore.byType.get( activeLinkType ).add( this );
        } else if ( storeHasType ) {
            this.componentStore.byType.get( activeLinkType ).delete( this );
        }
    }

    update() {
        super.update( ...arguments );
        this._registryType();
    }

    click( e ) {
        e.preventDefault();
        this.router.navigate( this.aNode.getAttribute( 'href' ) );
    }
}
