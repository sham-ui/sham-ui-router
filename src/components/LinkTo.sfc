<template>
    <a
            href={{this.generateURL( path, params )}}
            class={{this.generateClassName( className, useActiveClass, activeClass, path, params )}}
            :onclick={{::this.click}}
            :ref="aNode"
    >
        {{text}}
        {% defblock %}
    </a>
</template>

<script>
    import { inject, options } from 'sham-ui';
    import { onclick, ref } from 'sham-ui-directives';

    export default class LinkToComponent extends Template {
        constructor( options ) {
            super( {
                ...options,
                directives: {
                    onclick,
                    ref,
                    ...options.directives
                }
            } );
        }

        @inject router;
        @inject( 'sham-ui:store' ) componentStore;

        @options text = '';
        @options useActiveClass = false;
        @options activeClass = 'active';
        @options className = '';

        @options get params() {
            return {};
        }

        generateURL( path, params ) {
            return this.lastGeneratedURL = this.router.generate( path, params );
        }

        // eslint-disable-next-line no-unused-vars
        isActive( path, params ) {
            return this.lastGeneratedURL === this.router.lastRouteResolved().url;
        }

        generateClassName( className, useActiveClass, activeClass, path, params ) {
            return [
                className,
                useActiveClass && this.isActive( path, params ) ? activeClass : ''
            ].join( ' ' ).trim();
        }

        _registryType() {
            const activeLinkType = ACTIVE_PAGE_LINK_TYPE;
            const storeHasType = this.componentStore.byType.has( activeLinkType );
            if ( this.options.useActiveClass ) {
                if ( !storeHasType ) {
                    this.componentStore.byType.set( activeLinkType, new Set() );
                }
                this.componentStore.byType.get( activeLinkType ).add( this );
            } else if ( storeHasType ) {
                this.componentStore.byType.get( activeLinkType ).delete( this );
            }
        }

        update() {
            super.update( ...arguments );
            this._registryType();
        }

        click( e ) {
            e.preventDefault();
            this.router.navigate( this.aNode.getAttribute( 'href' ) );
        }
    }
</script>