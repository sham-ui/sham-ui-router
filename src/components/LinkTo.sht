<a
    href={{this.generateURL( path, params )}}
    class={{this.generateClassName( className, useActiveClass, activeClass, path, params )}}
    :onclick={{::this.click}}
    :ref="aNode"
>
    {{text}}
    {% defblock %}
</a>