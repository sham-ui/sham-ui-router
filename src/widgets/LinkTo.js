import { inject, options } from 'sham-ui';
import { onclick, ref } from 'sham-ui-directives';
import LinkToWidgetTemplate from './LinkTo.sht';
import { ACTIVE_PAGE_LINK_TYPE } from '../settings';

export default class LinkToWidget extends LinkToWidgetTemplate {
    constructor( options ) {
        super( {
            ...options,
            directives: {
                onclick,
                ref,
                ...options.directives
            }
        } );
    }

    @inject router;
    @inject( 'sham-ui:store' ) widgetStore;

    @options text = '';
    @options useActiveClass = false;
    @options activeClass = 'active';
    @options className = '';

    @options get params() {
        return {};
    }

    generateURL( path, params ) {
        return this.lastGeneratedURL = this.router.generate( path, params );
    }

    // eslint-disable-next-line no-unused-vars
    isActive( path, params ) {
        return this.lastGeneratedURL === this.router.lastRouteResolved().url;
    }

    generateClassName( className, useActiveClass, activeClass, path, params ) {
        return [
            className,
            useActiveClass && this.isActive( path, params ) ? activeClass : ''
        ].join( ' ' ).trim();
    }

    _registryType() {
        const storeHasType = this.widgetStore.byType.has( ACTIVE_PAGE_LINK_TYPE );
        if ( this.options.useActiveClass ) {
            if ( !storeHasType ) {
                this.widgetStore.byType.set( ACTIVE_PAGE_LINK_TYPE, new Set() );
            }
            this.widgetStore.byType.get( ACTIVE_PAGE_LINK_TYPE ).add( this );
        } else if ( storeHasType ) {
            this.widgetStore.byType.get( ACTIVE_PAGE_LINK_TYPE ).delete( this );
        }
    }

    update() {
        super.update( ...arguments );
        this._registryType();
    }

    click( e ) {
        e.preventDefault();
        this.router.navigate( this.aNode.getAttribute( 'href' ) );
    }
}
