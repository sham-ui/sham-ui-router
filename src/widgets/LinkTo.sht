<a
    href={{this.generateURL( path, params )}}
    class={{this.generateClassName( className, useActiveClass, activeClass, path, params )}}
    :onclick={{this.click.bind( this )}}
    :ref="aNode"
>
    {{text}}
    {% defblock %}
</a>