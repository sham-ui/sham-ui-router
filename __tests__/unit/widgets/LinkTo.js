import { DI } from 'sham-ui';
import LinkTo from '../../../src/widgets/LinkTo';
import renderer from 'sham-ui-test-helpers';

afterEach( () => {
    DI.bind( 'router', null );
} );

it( 'render correctly', () => {
    const generateMock = jest.fn();
    DI.bind( 'router', {
        generate: generateMock
    } );

    generateMock.mockReturnValueOnce( '/base' );

    const meta = renderer( LinkTo, {
        path: 'base',
        text: 'Base page'
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
    expect( generateMock.mock.calls.length ).toBe( 1 );
    expect( generateMock.mock.calls[ 0 ] ).toEqual( [ 'base', {} ] );
} );

it( 'params options', () => {
    const generateMock = jest.fn();
    DI.bind( 'router', {
        generate: generateMock
    } );

    generateMock.mockReturnValueOnce( '/base/1' );

    const meta = renderer( LinkTo, {
        path: 'base',
        params: {
            id: 1
        }
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
    expect( generateMock.mock.calls.length ).toBe( 1 );
    expect( generateMock.mock.calls[ 0 ] ).toEqual( [ 'base', { id: 1 } ] );
} );

it( 'useActiveClass options', () => {
    const generateMock = jest.fn();
    const lastRouteResolvedMock = jest.fn();

    DI.bind( 'router', {
        generate: generateMock,
        lastRouteResolved: lastRouteResolvedMock
    } );

    generateMock.mockReturnValue( '/base' );
    lastRouteResolvedMock.mockReturnValue( { url: '/base' } );

    const meta = renderer( LinkTo, {
        path: 'base',
        useActiveClass: true
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
    expect( generateMock.mock.calls.length ).toBe( 1 );
    expect( generateMock.mock.calls[ 0 ] ).toEqual( [ 'base', {} ] );
    expect( lastRouteResolvedMock.mock.calls.length ).toBe( 1 );

    DI.resolve( 'sham-ui' ).render.ALL();
} );

it( 'activeClass options', () => {
    const generateMock = jest.fn();
    const lastRouteResolvedMock = jest.fn();
    const registerActivePageLinkMock = jest.fn();

    DI.bind( 'router', {
        generate: generateMock,
        lastRouteResolved: lastRouteResolvedMock,
        _registerActivePageLink: registerActivePageLinkMock
    } );

    generateMock.mockReturnValueOnce( '/base' );
    lastRouteResolvedMock.mockReturnValueOnce( { url: '/base' } );

    const meta = renderer( LinkTo, {
        path: 'base',
        useActiveClass: true,
        activeClass: 'test-active'
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'className option', () => {
    const generateMock = jest.fn();
    DI.bind( 'router', {
        generate: generateMock
    } );

    generateMock.mockReturnValueOnce( '/base' );

    const meta = renderer( LinkTo, {
        path: 'base',
        text: 'Base page',
        className: 'foo bar'
    } );
    expect( meta.widget.querySelector( 'a' ).className ).toEqual( 'foo bar' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'className option & useActiveClass & activeClass', () => {
    const generateMock = jest.fn();
    const lastRouteResolvedMock = jest.fn();
    const registerActivePageLinkMock = jest.fn();

    DI.bind( 'router', {
        generate: generateMock,
        lastRouteResolved: lastRouteResolvedMock,
        _registerActivePageLink: registerActivePageLinkMock
    } );

    generateMock.mockReturnValueOnce( '/base' );
    lastRouteResolvedMock.mockReturnValueOnce( { url: '/base' } );

    const meta = renderer( LinkTo, {
        path: 'base',
        useActiveClass: true,
        activeClass: 'test-active',
        className: 'foo'
    } );
    expect( meta.widget.querySelector( 'a' ).className ).toEqual( 'foo test-active' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'click', () => {
    const generateMock = jest.fn();
    const navigateMock = jest.fn();

    DI.bind( 'router', {
        generate: generateMock,
        navigate: navigateMock
    } );

    generateMock.mockReturnValueOnce( '/base' );

    const meta = renderer( LinkTo, {
        path: 'base'
    } );
    meta.widget.querySelector( 'a' ).click();

    expect( meta.toJSON() ).toMatchSnapshot();
    expect( navigateMock.mock.calls.length ).toBe( 1 );
    expect( navigateMock.mock.calls[ 0 ] ).toEqual( [ '/base' ] );
} );

it( 'remove', () => {
    const generateMock = jest.fn();
    DI.bind( 'router', {
        generate: generateMock
    } );

    generateMock.mockReturnValueOnce( '/base' );

    const meta = renderer( LinkTo, {
        path: 'base',
        text: 'Base page'
    } );
    meta.widget.remove();
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'destroy', () => {
    const generateMock = jest.fn();
    DI.bind( 'router', {
        generate: generateMock
    } );

    generateMock.mockReturnValue( '/base' );

    const meta = renderer( LinkTo, {
        path: 'base',
        text: 'Base page'
    } );

    DI.resolve( 'sham-ui' ).render.ALL();

    expect( meta.toJSON() ).toMatchSnapshot();
    expect( generateMock.mock.calls.length ).toBe( 2 );
} );
