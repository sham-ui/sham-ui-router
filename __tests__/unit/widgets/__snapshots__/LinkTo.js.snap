// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activeClass options 1`] = `
Object {
  "Constructor": "LinkToWidget",
  "Options": Object {
    "ID": "widget",
    "activeClass": "test-active",
    "className": "",
    "containerSelector": "body",
    "directives": Object {
      "onclick": [Function],
      "ref": [Function],
    },
    "params": Object {},
    "path": "base",
    "text": "",
    "types": Array [],
    "useActiveClass": true,
  },
  "html": "
<a href=\\"/base\\" class=\\"test-active\\">
  <!--default-->
</a>
",
}
`;

exports[`className option & useActiveClass & activeClass 1`] = `
Object {
  "Constructor": "LinkToWidget",
  "Options": Object {
    "ID": "widget",
    "activeClass": "test-active",
    "className": "foo",
    "containerSelector": "body",
    "directives": Object {
      "onclick": [Function],
      "ref": [Function],
    },
    "params": Object {},
    "path": "base",
    "text": "",
    "types": Array [],
    "useActiveClass": true,
  },
  "html": "
<a href=\\"/base\\" class=\\"foo test-active\\">
  <!--default-->
</a>
",
}
`;

exports[`className option 1`] = `
Object {
  "Constructor": "LinkToWidget",
  "Options": Object {
    "ID": "widget",
    "activeClass": "active",
    "className": "foo bar",
    "containerSelector": "body",
    "directives": Object {
      "onclick": [Function],
      "ref": [Function],
    },
    "params": Object {},
    "path": "base",
    "text": "Base page",
    "types": Array [],
    "useActiveClass": false,
  },
  "html": "
<a href=\\"/base\\" class=\\"foo bar\\">Base page
  <!--default-->
</a>
",
}
`;

exports[`click 1`] = `
Object {
  "Constructor": "LinkToWidget",
  "Options": Object {
    "ID": "widget",
    "activeClass": "active",
    "className": "",
    "containerSelector": "body",
    "directives": Object {
      "onclick": [Function],
      "ref": [Function],
    },
    "params": Object {},
    "path": "base",
    "text": "",
    "types": Array [],
    "useActiveClass": false,
  },
  "html": "
<a href=\\"/base\\" class=\\"\\">
  <!--default-->
</a>
",
}
`;

exports[`destroy 1`] = `
Object {
  "Constructor": "LinkToWidget",
  "Options": Object {
    "ID": "widget",
    "activeClass": "active",
    "className": "",
    "containerSelector": "body",
    "directives": Object {
      "onclick": [Function],
      "ref": [Function],
    },
    "params": Object {},
    "path": "base",
    "text": "Base page",
    "types": Array [],
    "useActiveClass": false,
  },
  "html": "
<a href=\\"/base\\" class=\\"\\">Base page
  <!--default-->
</a>
",
}
`;

exports[`params options 1`] = `
Object {
  "Constructor": "LinkToWidget",
  "Options": Object {
    "ID": "widget",
    "activeClass": "active",
    "className": "",
    "containerSelector": "body",
    "directives": Object {
      "onclick": [Function],
      "ref": [Function],
    },
    "params": Object {
      "id": 1,
    },
    "path": "base",
    "text": "",
    "types": Array [],
    "useActiveClass": false,
  },
  "html": "
<a href=\\"/base/1\\" class=\\"\\">
  <!--default-->
</a>
",
}
`;

exports[`remove 1`] = `
Object {
  "Constructor": "LinkToWidget",
  "Options": Object {
    "ID": "widget",
    "activeClass": "active",
    "className": "",
    "containerSelector": "body",
    "directives": Object {
      "onclick": [Function],
      "ref": [Function],
    },
    "params": Object {},
    "path": "base",
    "text": "Base page",
    "types": Array [],
    "useActiveClass": false,
  },
  "html": "",
}
`;

exports[`render correctly 1`] = `
Object {
  "Constructor": "LinkToWidget",
  "Options": Object {
    "ID": "widget",
    "activeClass": "active",
    "className": "",
    "containerSelector": "body",
    "directives": Object {
      "onclick": [Function],
      "ref": [Function],
    },
    "params": Object {},
    "path": "base",
    "text": "Base page",
    "types": Array [],
    "useActiveClass": false,
  },
  "html": "
<a href=\\"/base\\" class=\\"\\">Base page
  <!--default-->
</a>
",
}
`;

exports[`useActiveClass options 1`] = `
Object {
  "Constructor": "LinkToWidget",
  "Options": Object {
    "ID": "widget",
    "activeClass": "active",
    "className": "",
    "containerSelector": "body",
    "directives": Object {
      "onclick": [Function],
      "ref": [Function],
    },
    "params": Object {},
    "path": "base",
    "text": "",
    "types": Array [],
    "useActiveClass": true,
  },
  "html": "
<a href=\\"/base\\" class=\\"active\\">
  <!--default-->
</a>
",
}
`;
